### DISCUSSION TITLE
Turn Based Patterns and Ideas using Bevy ECS

### DISCUSSION THREAD

**Discussion by mirkoRainer** [bevyengine/bevy#3370] [Ideas]
Considering a complex turn-based game. 

What are some good patterns/practices when using Bevy ECS?

Ideas derived from a brief read of a Discord thread:
- One monolithic system to handle most things
- Several systems triggering from events
- Less emphasis on fast iteration using ECS but heavy use of queries (possibly in a sequential order)
- Use of sparse set components
- Use of system scheduling algorithm

**Comment by nicopap** [comment]
Turns out I am making a hybrid turn based game in bevy. This may be surprising, but I've had no issues with modelling it in the ECS. 

My approach consists of having two states "Planning" (where the player actually choses the action for the game to run) and "Running" (where the actions are being played)

Typically the World variables for the "Running" state are immediately set as soon as the player decides to end his turn. Then, all the setup systems for the "Running" state are ran. You will probably need to set an execution order on those, because it is likely that you will need outputs of some systems as input of others.

The World once setup, is available to all the real time systems that do game simulation and animations (like moving pieces on a chessboard). I have an always running system that detects if all the Running simulations and animations are done and then switch back the game state to "Planning". The loop then repeats.

**Reply by nicopap** [reply]
I'm a bit ashamed of the way I check if all systems are done. All relevant systems for the Running state are modeled using Hierarchical finite state machines. (hfsm) plug for my library btw: https://github.com/nicopap/pure-hfsm

In all of my state machines, I have the equivalent of a "Planning" machine that transitions into "Running" automatically when the game state changes. in hfsm, this corresponds to an additional item in the machine stack. The "Running" machines work on timers or anything relevant, and once we reach the exit state, the "Running" machine is popped, removed from the stack. The `check if over` system basically is just a
```rust
if hfsm_query.iter().all(|hfsm| hfsm.stack_len() == hfsm.planning_stack_len) {
    state.switch_to(GameState::Planning);
}
```
Ideally I would check the actual state and machine rather than the stack len, but it works for now…
